File: src/imbalance.ts
/**
 * Book and trade imbalance metrics.
 *
 * Both return a value in [-1, 1]: positive = buy-side heavy, negative =
 * sell-side heavy, 0 = balanced (or empty input).
 */
import type { L1Quote, Trade } from "./types.ts";

/** Top-of-book depth imbalance: (bidSize − askSize) / (bidSize + askSize). */
export function depthImbalance(q: L1Quote): number {
  const denom = q.bidSize + q.askSize;
  return denom === 0 ? 0 : (q.bidSize - q.askSize) / denom;
}

/** Trade imbalance: (buyVolume − sellVolume) / (buyVolume + sellVolume). */
export function tradeImbalance(trades: Trade[]): number {
  let buy = 0;
  let sell = 0;
  for (const t of trades) {
    if (t.side === "buy") buy += t.size;
    else sell += t.size;
  }
  const denom = buy + sell;
  return denom === 0 ? 0 : (buy - sell) / denom;
}

File: python/src/orderflow_metrics/imbalance.py
"""Book and trade imbalance metrics.

Both return a value in [-1, 1]: positive = buy-side heavy, negative = sell-side
heavy, 0 = balanced (or empty input).
"""
from __future__ import annotations

from typing import Sequence

from .types import L1Quote, Trade


def depth_imbalance(q: L1Quote) -> float:
    """Top-of-book depth imbalance: (bid_size - ask_size) / (bid_size + ask_size)."""
    denom = q.bid_size + q.ask_size
    return 0.0 if denom == 0 else (q.bid_size - q.ask_size) / denom


def trade_imbalance(trades: Sequence[Trade]) -> float:
    """Trade imbalance: (buy_vol - sell_vol) / (buy_vol + sell_vol)."""
    buy = sum(t.size for t in trades if t.side == "buy")
    sell = sum(t.size for t in trades if t.side == "sell")
    denom = buy + sell
    return 0.0 if denom == 0 else (buy - sell) / denom
